### DISCUSSION TITLE
Semantic changes in the model definition

### DISCUSSION THREAD

**Discussion by Filippo-Tamagnini** [do-mpc/do-mpc#135] [Ideas]
**1) Measurements should be outputs**
When building a model of the form xdot = f(x,u); y=g(x,u) (i'm simplifying things here), y is a generic output of the system to the environment. This quantity could then be measured or not, but this should pertain to the definition of the observer.
Example: a sensor provides an output, but it is only reliable in a certain region of the state space, thus making it a measurement only in a limited set of cases
**2) Unique function for setting expressions**
Currently, separate functions exist to define expressions for the ode, the measurements/outputs, the algebraic equations and the auxiliary expressions. I think they could all be safely replaced with a one-size-fits-all set_rhs(var_type, var_name, expression, process_noise), but i realize this could be a problem for the algebraic states (setting an algebraic expression for an algebraic state could lead users to think that only explicit algebraic expressions are supported) and would ultimately break the existing code. 
**3) Time-varying parameters should be parameters**
From what i have seen, the difference between _p and _tvp is mainly in the notation. The set_tvp_fun and set_p_fun methods are however defined in separate places (set_tvp_fun is actually defined twice: for the controller and for the estimator).
I think the code would be simplified by condensing the two functions in one, and perhaps setting the time-varyingness of the parameter as a property of the parameter

**Comment by 4flixt** [comment]
Dear Filippo,

thank you for your comments! Here is my take:

1) I am not entirely sure what you mean here but I guess you are correct that what we refer to as "measurements" are "outputs" of the system. They only become measurements after being processed by a sensor block (which typically only adds noise). We bake both process and measurement noise in the model as this significantly facilitates using it as the base of a MHE formulation. I am not sure what to take from your example. What is your proposed change regarding this scenario?

2) The right-hand-side (f) the measurement equations (g) and the algebraic equations are fundamentally different. Somehow the user has to indicate which of these functions he/she is currently setting. Having different methods for all of them (with individual doc string) is one way of doing this. Alternatively, your one-size-fits-all function would require additional parameters (a "flag") to make this indication. This would make life more difficult imo, as e.g. the ``set_rhs`` method needs a parameter indicating for which state the function is being set. The ``set_meas`` doesn't need this parameter.

3) Time-varying parameters and parameters are fundamentally different. Please see this [FAQ article](https://www.do-mpc.com/en/latest/FAQ.html#time-varying-parameters) on the purpose of TVP (and the difference to parameters).

Finally, I will move this issue to the discussion section of the repository and open individual issues as points-of-actions arise from this discussion.

Thanks again for the feedback! 

4flixt

**Comment by Filippo-Tamagnini** [comment]
1) When plugging the measurement structure into the MHE or any other estimator, we could allow NaN values (measurements not available), and then compute the estimator cost based on the available measurements only. This would address the point raised in https://github.com/do-mpc/do-mpc/discussions/134
How to do this in practice: the cost is based on the weighted norm ||y{pred} - y{meas}|| with respect to some weight matrix W. If a measurement is not availbable, the corresponding weight could be set to zero by pre and post multiplying W by some masking matrix M (i'm thinking of an identity matrix with the diagonal elements corresponding to unavailable measurements set to zero).
This could also extend the functionality of the MHE in presence of delayed measurements
2) No objections to that
3) I'll have a deeper look

**Comment by 4flixt** [comment]
1) We could do something like this. The current approach is quite simple, however: Just introduce a ``_tvp``  (the weight matrix you mentioned). The user can then flexibly use this weight in the stage-cost and e.g. matching of measured (not available at all steps) and computed measurement. This does take some work from the side of the user but again: It is super flexible and the user can fully understand what is going on. This wouldn't be the case if we do something in the backend which the users can't see (even if it is slightly more convenient).

**Comment by Filippo-Tamagnini** [comment]
I see your point, however the tvp_fun() function only depends on time. Does this mean that the user should redefine it whenever a measurement becomes unavailable for any non-deterministic reason?
